design/one-pager-component-mutable-and-versioning.md

```diff
@@ -0,0 +1,340 @@
+metadata:
+  labels:
+    controller-revision-hash: c8bb659c5
+  name: frontend-c8bb659c5
+  namespace: default
+  ownerReferences:
+  - apiVersion: core.oam.dev/v1alpha2
+    blockOwnerDeletion: true
+    controller: true
+    kind: Component
+    name: frontend
+revision: 1
+data:
+  spec:
+    workload:
+      apiVersion: core.oam.dev/v1alpha2
+      kind: ContainerizedWorkload
+      spec:
+        containers:
+        - name: my-cool-workload
+          image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+          cmd:
+          - "bash lscpu"
+```
+
+Make a change to `component.yaml`:
+
+```shell script
+           cmd:
+-            - "bash lscpu"
++            - "bash top"
+```
+
+A new `ControllerRevision` will be automatically generated:
+
+```shell script
+$ kubectl get controllerrevisions
+NAMESPACE     NAME                  CONTROLLER               REVISION   AGE
+default       frontend-c8bb659c5    core.oam.dev/component   1          2d15h
+default       frontend-a75588698    core.oam.dev/component   2          2d14h
+
+$ kubectl get controllerrevision frontend-a75588698 -o yaml
+apiVersion: apps/v1
+kind: ControllerRevision
+metadata:
+  labels:
+    controller-revision-hash: a75588698
+  name: frontend-a75588698
+  namespace: default
+  ownerReferences:
+  - apiVersion: core.oam.dev/v1alpha2
+    blockOwnerDeletion: true
+    controller: true
+    kind: Component
+    name: frontend
+revision: 2
+data:
+  spec:
+    workload:
+      apiVersion: core.oam.dev/v1alpha2
+      kind: ContainerizedWorkload
+      spec:
+        containers:
+        - name: my-cool-workload
+          image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+          cmd:
+          - "bash top"
+```
+
+Now that the basic controller revision are in place, we can use them in Trait and ApplicationConfiguration.
+
+### A real running workload instance will have the same name with revision
+
+I we create an applicationConfiguration like below:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: frontend
+```
+
+The real workload instance is created with the same name of the controllerRevision.
+
+```yaml
+  apiVersion: core.oam.dev/v1alpha2
+  kind: ContainerizedWorkload
+  metadata:
+    name: frontend-a75588698
+  spec:
+    containers:
+    - name: my-cool-workload
+      image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+      cmd:
+      - "bash lscpu"
+```
+
+
+### ControllerRevision could be referenced by Trait 
+
+Some traits can be revision aware who can specify a component revision and won't be affected
+by the change of component.
+
+For these traits, they have to declare they are revision aware in TraitDefinition like below:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  name: fancytraits.core.oam.dev
+spec:
+  revision:
+    aware: true
```

Comment:
I wonder when do we still  need a  `spec.revision` field when `spec.revision.aware` is false?

Reply:
I changed to alwaysControl, and this will have new meanings: